### ISSUE TITLE
negative value in smoothed_forecasts_error_cov in Kalman Filter

### ISSUE THREAD

**Issue by lithces** [statsmodels/statsmodels#5044]
I got negative value in statsmodels.tsa.statespace.kalman_smoother.SmootherResults.smoothed_forecasts_error_cov. 

Is it expected or there is something wrong with the computation logic?

I am using  a custom SSM. The code to reproduce the output is as follows,

```python
class LocalLinearTrendSeasonal(sm.tsa.statespace.MLEModel):
    def __init__(self, endog, s=7):
        # Model order
        if s<=1:
            raise ValueError('s should be greater than 1')

            
        k_states = k_posdef = 2 + s - 1
        k_posdef = 3

        # Initialize the statespace
        super(LocalLinearTrendSeasonal, self).__init__(
            endog, k_states=k_states, k_posdef=k_posdef,
            initialization='approximate_diffuse',
            loglikelihood_burn=k_states
        )

        # Initialize the matrices
        tmp = np.zeros(k_states)
        tmp[0:3] = np.array([1, 0, 1])
        self.ssm['design'] = tmp
        tras1 = np.array([[1, 1], [0, 1]])
        tras2 = np.diag(np.ones(s-2), k=-1)
        tras2[0,:] = -1
        
        self.ssm['transition'] = np.block([ [tras1, np.zeros( (2, s-1))], [np.zeros( (s-1,2)), tras2]])        
        self.ssm['selection'] = np.vstack([np.eye(3), np.zeros( (s-2, 3))])

        # Cache some indices
        self._state_cov_idx = ('state_cov',) + np.diag_indices(k_posdef)

    @property
    def param_names(self):
        return ['sigma2.measurement', 'sigma2.level', 'sigma2.trend', 'sigma2.seasonality']

    @property
    def start_params(self):
        return [np.std(self.endog)]*4

    def transform_params(self, unconstrained):
        return unconstrained**2

    def untransform_params(self, constrained):
        return constrained**0.5

    def update(self, params, *args, **kwargs):
        params = super(LocalLinearTrendSeasonal, self).update(params, *args, **kwargs)
        
        # Observation covariance
        self.ssm['obs_cov',0,0] = params[0]

        # State covariance
        self.ssm[self._state_cov_idx] = params[1:]
        
        

y_in = np.array([0.5129303 , 0.47890179, 0.47064982, 0.42271865, 0.4529535 ,
       0.38761738, 0.38819515, 0.41629387, 0.50301191, 0.56177213,
       0.55244079, 0.51377614, 0.44047094, 0.44356177, 0.49939891,
       0.53397997, 0.55713189, 0.55572101, 0.57910139, 0.54201371,
       0.53576324, 0.54977693, 0.62479045, 0.62152558, 0.56015817,
       0.50991228, 0.38478678, 0.34842118, 0.36945561, 0.33365732,
       0.3362659 , 0.32574144, 0.32315578, 0.33374874, 0.34861643,
       0.4086186 , 0.42925111, 0.49206309, 0.53846398, 0.55311584,
       0.5389626 , 0.54129383, 0.56922575, 0.61613736, 0.65956821,
       0.71320605, 0.72123158, 0.67122813, 0.66344295, 0.65256848])
mod = LocalLinearTrendSeasonal(y_in)
res = mod.fit(disp=False)  
y_smooth_cov = res.smoother_results.smoothed_forecasts_error_cov.squeeze()
sum(y_smooth_cov<0)
```

**Comment by lithces** [comment]
Thanks @ChadFulton for the quick reply. I took your advise and it works for my problem.

Is there any document describing how the SSM inference is made in statsmodel？I am a little bit confused about the initialization. In my opinion, assuming all Gaussian noise, given the observation, input, and hyper-parameters (sigma2* in my example), the posterior of latent variables (states given all the observations) is a multivariate normal distribution, and I am not able to tell if initialization has any connection with the posterior.

**Comment by ChadFulton** [comment]
Yes, the book "Time Series Analysis by State Space Methods" by Durbin and Koopman (2010).

Your statement is generally true. The algorithm is recursive, and initialization refers to the assumed prior distribution of states. Approximate diffuse initialization sets the prior to N(0, kappa) where kappa is some large number (e.g. 10^6) - this is an approximate technique that can sometimes lead to some numerical artifacts, as you've seen in your example, and this is why it is suggested that you ignore the output associated the first `k` observations. Exact diffuse initialization takes the limit as kappa -> infinity, and does not have those numerical problems.

**Comment by ChadFulton** [comment]
Closing since I don't think this is a bug, but feel free to continue the discussion.

**Comment by lithces** [comment]
> Yes, the book "Time Series Analysis by State Space Methods" by Durbin and Koopman (2010).

Sorry for the late reply. 

Unfortunately I have no access to the book :(.  After some search, I found a paper (learning stable dynamical systems)[https://www.ml.cmu.edu/research/dap-papers/dap_boots.pdf] describing two parameter estimation methods. Is statsmodels using the EM method described in Sec. 3.1?

Thanks!

**Comment by ChadFulton** [comment]
We don't use the EM algorithm - instead, we maximize the loglikelihood directly using numerical techniques.

Of course the smoothed states are available to you, so you could implement the EM algorithm if you like.

**Comment by lithces** [comment]
> We don't use the EM algorithm - instead, we maximize the loglikelihood directly using numerical techniques.

Just found [pykalman](https://github.com/pykalman/pykalman) in which EM is implemented. However it takes so many iterations to converge.

I tried to implement a maximum likelihood algorithms (states are marginalized out analytically, and the optimization is with the missing parameters) with Pytorch. I tried Adam and LBFGS, however the log likelihood is always worse then the implementation in statsmodels. Sometimes LBFGS leads to nan... 

Do you have tips on the implementation? e.g. which algorithms should be used, how to initialize, how to prevent from numerical instability?

**Comment by josef-pkt** [comment]
I never tried to understand the details of the kalman filter implementation, but I'm looking currently at Ridge penalization and regularization that has somewhat similar features:

general possibilities to improve robustness
- use an informative prior or initial state that forces the model to behave nicely even in small samples
- use regularization on covariance, i.e. force it to be positive definite or positive semi-definite, either by adding a Ridge factor or by clipping eigenvalues that are too small or negative.
- maybe force the model to be in the interior of the parameter space.

(My current context: L2 penalized least squares and GAM where either the prior covariance, i.e. penalty matrix, or the hessian of the log-likelihood are in general singular)
